# Data Visualization (Trực quan hóa dữ liệu)

Ước tính thời gian cần thiết: **30** phút

## Mục tiêu

Sau khi hoàn thành bài Lab này, bạn sẽ có thể:

-   Tạo visualization của dữ liệu bằng Python
-   Sử dụng các thư viện Python khác nhau để tạo visualization


## Mục lục

<div class="alert alert-block alert-info" style="margin-top: 20px">

1.  [Explore tập dữ liệu với _pandas_](#0)<br>
    1.1 [Tập dữ liệu: Nhập cư đến Canada từ 1980 đến 2013](#2)<br>
    1.2 [Những điều cơ bản về _pandas_](#4) <br>
    1.3 [_pandas_ Intermediate: Tạo chỉ mục và chọn Indexing and Selection](#6) <br>
2.  [Visualize dữ liệu bằng Matplotlib](#8) <br>
    2.1 [Matplotlib: Thư viện Visualization tiêu chuẩn của Python](#10) <br>
3.  [Biểu đồ đường](#12)
    </div>
    <hr>


# Explore tập dữ liệu với _pandas_ <a id="0"></a>

_pandas_ là một bộ công cụ phân tích dữ liệu rất cần thiết cho Python. Từ [website](http://pandas.pydata.org?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ) của họ:

> _pandas_ là một gói Python cung cấp cấu trúc dữ liệu nhanh, linh hoạt và có tính biểu đạt cao được thiết kế để giúp làm việc với dữ liệu “quan hệ” hoặc “có nhãn” dễ dàng và trực quan. Mục đích của nó là trở thành khối xây dựng cấp cao cơ bản để thực hiện phân tích dữ liệu ** thực tế ** bằng Python.

Khóa học chủ yếu dựa vào _pandas_ để xử lý dữ liệu (data wrangling), phân tích và để visualize. Bạn nên dành chút thời gian tự làm quen với Tham chiếu API _pandas_: [http://pandas.pydata.org/pandas-docs/stable/api.html](http://pandas.pydata.org/pandas-docs/stable/api.html?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).


## Tập dữ liệu: Nhập cư đến Canada từ 1980 đến 2013 <a id="2"></a>


Nguồn của tập dữ liệu: [Luồng nhập cư quốc tế đến và đi từ các quốc gia được chọn - Bản sửa đổi năm 2015](http://www.un.org/en/development/desa/population/migration/data/empirical2/migrationflows.shtml?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).

Tập dữ liệu chứa dữ liệu hàng năm về dòng người nhập cư quốc tế được các quốc gia đích ghi lại. Dữ liệu thể hiện cả dòng người nhập cư và dòng người di cư theo các dữ liệu như place of birth (nơi sinh), citizenship (quốc tịch) hoặc place of previous / next residence (nơi cư trú trước đây / tiếp theo) của cả người nước ngoài và người dân mang quốc tịch. Phiên bản này trình bày dữ liệu liên quan đến 45 quốc gia.

Trong bài Lab này, chúng ta sẽ tập trung vào dữ liệu nhập cư Canada.

<img src = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png" align="center" width=900>

Để các bạn dễ dàng hoàn thành bài Lab hơn, dữ liệu nhập cư của Canada đã được trích xuất và tải lên một server của IBM. Bạn có thể tìm nạp dữ liệu ở [đây](https://ibm.box.com/shared/static/lw190pt9zpy5bd1ptyg2aw15awomz9pu.xlsx?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).

* * *


## Những điều cơ bản về _pandas_<a id="4"></a>


Đầu tiên, chúng ta impotrt hai module phân tích dữ liệu chính: _pandas_ và **Numpy**.


In [76]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library
!conda install -c anaconda xlrd --yes

/bin/bash: line 1: conda: command not found


Hãy download và import tập dữ liệu Nhập cư Canada chính bằng phương thức _pandas_ `read_excel()`. Thông thường, trước khi có thể làm điều đó thì chúng ta sẽ cần download một module mà _pandas_ cần để đọc file excel. Module này là **xlrd**. Để giúp bạn làm bài thuận tiện hơn, chúng tôi đã cài đặt module này trước cho bạn, nên bạn không cần lo về điều đó nhé. Nếu không thì bạn sẽ phải cho chạy dòng code dưới đây để cài đặt module **xlrd**:

```
!conda install -c anaconda xlrd --yes
```


Bây giờ chúng ta đã sẵn sàng để đọc dữ liệu.


In [91]:
df_can = pd.read_excel('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
                        sheet_name='Canada by Citizenship',
                       skiprows=range(20),
                       skipfooter=2)

print ('Data read into a pandas dataframe!')

Data read into a pandas dataframe!


Hãy xem 5 hàng trên cùng của tập dữ liệu bằng hàm `head ()`.


In [78]:
df_can.head()
# tip: You can specify the number of rows you'd like to see as follows: df_can.head(10)

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

Chúng ta cũng có thể xem 5 hàng dưới cùng của tập dữ liệu bằng hàm `tail ()`.


In [79]:
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

Khi phân tích một tập dữ liệu, bạn nên bắt đầu bằng cách lấy thông tin cơ bản về khung dữ liệu. Chúng ta có thể làm điều này bằng cách sử dụng phương thức `info ()`.

Có thể dùng phương thức này để có được một bản tóm tắt ngắn gọn về khung dữ liệu.


In [80]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


Để có danh sách các cột headers (tiêu đề), chúng ta có thể gọi tham số `.columns` của khung dữ liệu.


In [81]:
df_can.columns.values

array(['Type', 'Coverage', 'OdName', 'AREA', 'AreaName', 'REG', 'RegName',
       'DEV', 'DevName', 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987,
       1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998,
       1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
       2010, 2011, 2012, 2013], dtype=object)

Tương tự, để lấy danh sách các chỉ số, chúng ta sử dụng tham số `.index`.


In [82]:
df_can.index.values

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

Lưu ý: Kiểu chỉ mục và cột mặc định KHÔNG phải là danh sách.


In [83]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


Để lấy chỉ mục và các cột dưới dạng danh sách, chúng ta có thể sử dụng phương thức `tolist ()`.


In [84]:
df_can.columns.tolist()
df_can.index.tolist()

print (type(df_can.columns.tolist()))
print (type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


Để xem các chiều của khung dữ liệu, chúng ta dùng tham số `.shape`.


In [85]:
# size of dataframe (rows, columns)
df_can.shape

(195, 43)

Lưu ý: Các kiểu chính được lưu trữ trong các đối tượng _pandas_ là _float_, _int_, _bool_, _datetime64 [ns]_ và _datetime64 [ns, tz] \ (in> = 0.17.0)_, _timedelta [ns]_, _category (in> = 0,15.0)_ và _object_ (string). Ngoài ra, các dtype này có kích thước của các item, ví dụ: int64 và int32.



Để làm sạch tập dữ liệu để bỏ một vài cột không cần thiết, chúng ta có thể sử dụng phương thức _pandas_ `drop ()` như sau:


In [92]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

Hãy đổi tên các cột để chúng có ý nghĩa. Chúng ta có thể sử dụng phương thức `rename ()` bằng cách truyền từ điển gồm các tên cũ và mới như sau:

In [93]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

Chúng ta cũng sẽ thêm cột 'Total' để tính tổng số người nhập cư theo quốc gia trong toàn bộ giai đoạn 1980 - 2013, như sau:


In [96]:
#df_can['Total'] = df_can.sum(axis=1)
df_can["Total"] = df_can.select_dtypes(include=["int64"]).sum(axis=1)

Để kiểm tra xem có bao nhiêu đối tượng null trong tập dữ liệu, chúng ta có thể dùng cú pháp sau:


In [32]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
dtype: int64

Cuối cùng, hãy xem tóm tắt của từng cột trong khung dữ liệu bằng phương thức `description ()`.


In [97]:
df_can.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

       ...          2005          2006          2007          2008  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1320.292308   1266.958974   1191.820513   1246.394872   
std    ...   4425.957828   3926.717747   3443.542409   3694.573544   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.500000     25.000000     31.000000     31.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    832.000000    842.000000    899.000000    934.500000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1275.733333   1420.287179   1262.533333   1313.958974   1320.702564   
std     3829.630424   4462.946328   4030.084313   4247.555161   4237.951988   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       36.000000     40.500000     37.500000     42.500000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      888.000000    932.000000    772.000000    783.000000    796.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

               Total  
count     195.000000  
mean    32867.451282  
std     91785.498686  
min         1.000000  
25%       952.000000  
50%      5018.000000  
75%     22239.500000  
max    691904.000000  

[8 rows x 35 columns]

* * *

## _pandas_ Intermediate: Tạo chỉ mục và chọn (slicing - cắt)<a id="6"></a>


### Chọn cột

**Có hai cách để lọc tên cột:**

Phương pháp 1: Nhanh chóng và dễ dàng, nhưng chỉ hoạt động nếu tên cột KHÔNG có khoảng trắng hoặc ký tự đặc biệt.

```python
    df.column_name
        (returns series)
```

Phương pháp 2: Mạnh mẽ hơn và có thể lọc trên nhiều cột.

```python
    df['column']  
        (returns series)
```

```python
    df[['column 1', 'column 2']]
        (returns dataframe)
```

* * *


Example: Let's try filtering on the list of countries ('Country').


In [102]:
df_can.Country
# returns a series

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Let's try filtering on the list of countries ('OdName') and the data for years: 1980 - 1985.


In [103]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe
# notice that 'Country' is string, and the years are integers.
# for the sake of consistency, we will convert all column names to string later on.

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Chọn hàng

Có 3 cách chính để chọn hàng:

```python
    df.loc[label]        
        #filters by the labels of the index/column
    df.iloc[index]       
        #filters by the positions of the index/column
```


Trước khi chúng ta tiếp tục, hãy lưu ý rằng chỉ mục mặc định của tập dữ liệu là các số từ 0 đến 194. Điều này sẽ gây khó khăn cho việc thực hiện truy vấn của một quốc gia cụ thể. Ví dụ, để tìm kiếm dữ liệu về Nhật Bản, chúng ta cần biết giá trị chỉ mục tương ứng của nước này.

Chúng ta có thể khắc phục điều này rất dễ dàng bằng cách đặt cột 'Country' làm chỉ mục bằng phương thức `set_index ()`.


In [104]:
df_can.set_index('Country', inplace=True)
# tip: The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [65]:
df_can.head(3)

Continent           Region             DevName  1980  1981  1982  \
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2004  2005  2006  2007  2008  2009  \
Afghanistan    47    71   340   496  ...  2978  3436  3009  2652  2111  1746   
Albania         0     0     0     1  ...  1450  1223   856   702   560   716   
Algeria        69    63    44    69  ...  3616  3626  4807  3623  4005  5393   

             2010  2011  2012  2013  
Afghanistan  1758  2203  2635  2004  
Albania       561   539   620   603  
Algeria      4752  4325  3774  4331  

[3 rows x 37 columns]

In [105]:
# optional: to remove the name of the index
df_can.index.name = None

Ví dụ: Hãy xem số lượng người nhập cư từ Nhật Bản (hàng 87) cho các tình huống sau:

```
1. Toàn bộ dữ liệu của hàng (tất cả các cột)
2. Trong năm 2013
3. Từ năm 1980 đến 1985
```


In [106]:
# 1. the full row data (all columns)
print(df_can.loc['Japan'])

# alternate methods
print(df_can.iloc[87])
print(df_can[df_can.index == 'Japan'].T.squeeze())

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [75]:
# 2. for year 2013
# print(df_can.loc['Japan', 2013])
print(df_can.loc['Japan', 2013])


# alternate method
#print(df_can.iloc[87, 36]) # year 2013 is the last column, with a positional index of 36

KeyError: 2013

In [70]:
# 3. for years 1980 to 1985
print(df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]])
print(df_can.iloc[87, [3, 4, 5, 6, 7, 8]])

KeyError: "None of [Index([1980, 1981, 1982, 1983, 1984, 1984], dtype='int64')] are in the [index]"

Các tên cột là số nguyên (chẳng hạn như years) có thể gây nhầm lẫn. Ví dụ: khi chúng ta tham chiếu năm 2013, người ta có thể nhầm lẫn rằng đó là chỉ số vị trí thứ 2013.

Để tránh gây nhầm lẫn như vậy, hãy chuyển đổi tên cột thành string: '1980' thành '2013'.


In [ ]:
df_can.columns = list(map(str, df_can.columns))
# [print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

Vì chúng ta đã chuyển đổi years thành string, hãy khai báo một biến sẽ cho phép chúng ta dễ dàng gọi toàn bộ phạm vi năm:


In [ ]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

### Lọc dựa trên một tiêu chí

Để lọc khung dữ liệu dựa trên một điều kiện, chúng ta chỉ cần truyền điều kiện dưới dạng một vectơ boolean.

Ví dụ: Hãy lọc khung dữ liệu để hiển thị dữ liệu về các quốc gia Châu Á (AreaName = Asia).


In [ ]:
# 1. create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

In [ ]:
# 2. pass this condition into the dataFrame
df_can[condition]

In [ ]:
# we can pass mutliple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Trước khi tiếp tục: hãy xem lại chúng ta đã thực hiện những thay đổi gì với khung dữ liệu của mình.


In [ ]:
print('data dimensions:', df_can.shape)
print(df_can.columns)
df_can.head(2)

* * *

# Visualize dữ liệu bằng Matplotlib<a id="8"></a>


## Matplotlib: Thư viện Visualization tiêu chuẩn của Python<a id="10"></a>

Trong khóa học này, chúng ta sẽ chủ yếu tìm hiểu thư viện vẽ sơ đồ [Matplotlib](http://matplotlib.org?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).  Như trên website của họ đã nói:

> Matplotlib là một thư viện vẽ đồ thị Python 2D, nó tạo ra các đồ thị với chất lượng xuất bản ở nhiều định dạng bản cứng và môi trường tương tác trên nhiều nền tảng. Có thể dùng Matplotlib trong các tập lệnh Python, shell Python và IPython, jupyter notebook, server ứng dụng web và 4 bộ công cụ giao diện người dùng đồ họa.

Nếu bạn muốn tạo ra visualization tốt bằng python, Matplotlib là một công cụ cần thiết mà bạn có thể sử dụng.


### Matplotlib.Pyplot

Một trong những khía cạnh cốt lõi của Matplotlib là `matplotlib.pyplot`. Đó chính là scripting layer (giao diện người dùng sẽ thấy) của Matplotlib, chúng ta đã tìm hiểu chi tiết về nó trong các video về Matplotlib. Mọi người hãy nhớ lại rằng nó là một tập hợp các hàm kiểu lệnh khiến cho Matplotlib hoạt động giống như MATLAB. Mỗi hàm `pyplot` sẽ thực hiện một số thay đổi với một hình: ví dụ: tạo một hình, tạo vùng vẽ đồ thị trong một hình, vẽ một số đường trong vùng vẽ đồ thị, trang trí đồ thị bằng nhãn, v.v. Trong bài Lab này, chúng ta sẽ làm việc với scripting layer để tìm hiểu cách tạo các biểu đồ đường. Trong các bài Lab trong tương lai, chúng ta cũng sẽ làm việc với Artist layer để thử nghiệm trực tiếp xem nó khác với scritpting layer như thế nào.


Hãy bắt đầu bằng cách import `Matplotlib` và `Matplotlib.pyplot` như sau:


In [ ]:
# we are using the inline backend
%matplotlib inline

import matplotlib as mpl
import matplotlib.pyplot as plt

\*không bắt buộc: kiểm tra xem Matplotlib đã được tải chưa.


In [ ]:
print ('Matplotlib version: ', mpl.__version__) # >= 2.0.0

\*không bắt buộc: áp dụng một style cho Matplotlib.


In [ ]:
print(plt.style.available)
mpl.style.use(['ggplot']) # optional: for ggplot-like style

### Vẽ biểu đồ trong _pandas_

May mắn thay, chúng ta có thể dùng Matplotlib đã được cài sẵn trong pandas. Vẽ đồ thị trong _pandas_ cũng đơn giản như thêm phương thức `.plot ()` vào một chuỗi hoặc khung dữ liệu.

Tài liệu:

-   [Vẽ biểu đồ với chuỗi](http://pandas.pydata.org/pandas-docs/stable/api.html#plotting?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)<br>
-   [Vẽ biểu đồ với khung dữ liệu](http://pandas.pydata.org/pandas-docs/stable/api.html#api-dataframe-plotting?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork-20297740&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


# Biểu đồ đường (Chuỗi/Khung dữ liệu) <a id="12"></a>


**Biểu đồ đường là gì và tại sao chúng ta lại sử dụng nó?**

Biểu đồ đường hoặc đồ thị đường là một loại biểu đồ hiển thị thông tin dưới dạng một chuỗi các điểm dữ liệu được gọi là 'marker' được kết nối bằng các đoạn đường thẳng. Đây là một loại biểu đồ cơ bản rất phổ biến trong nhiều lĩnh vực.
Hãy sử dụng biểu đồ đường khi bạn có tập dữ liệu liên tục. Chúng phù hợp nhất cho việc visualize dữ liệu theo xu hướng trong một khoảng thời gian.


**Cùng bắt đầu với một case study nhé:**

Năm 2010, Haiti phải hứng chịu thảm họa động đất mạnh 7,0 độ richter. Trận động đất đã tàn phá và gây nhiều thiệt hại về người trên diện rộng và 3 triệu người đã bị ảnh hưởng bởi thảm họa thiên nhiên này. Canada đã rất nỗ lực để giúp đỡ và Chính phủ Canada đã tăng cường tiếp nhận người tị nạn từ Haiti. Chúng ta có thể visualize sự nỗ lực của họ rất nhanh chóng bằng cách sử dụng biểu đồ `Line`:

**Câu hỏi:** Vẽ biểu đồ đường của luồng di cư từ Haiti bằng `df.plot ()`.


Đầu tiên, chúng ta sẽ trích xuất chuỗi dữ liệu của Haiti.


In [ ]:
haiti = df_can.loc['Haiti', years] # passing in years 1980 - 2013 to exclude the 'total' column
haiti.head()

Tiếp theo, chúng ta sẽ vẽ một biểu đồ đường bằng cách thêm `.plot ()` vào khung dữ liệu `haiti`.


In [ ]:
haiti.plot()

_pandas_ tự động điền các giá trị chỉ số (year - năm) vào trục x và các giá trị cột (population - dân số) vào trục y. Tuy nhiên, các bạn cần lưu ý rằng các năm không được hiển thị vì chúng thuộc kiểu _string_ (xâu). Do đó, hãy thay đổi kiểu của giá trị chỉ mục thành _integer_ (số nguyên) để vẽ biểu đồ.

Ngoài ra, hãy gắn nhãn trục x và y bằng `plt.title()`, `plt.ylabel()`, và `plt.xlabel()` như sau:


In [ ]:
haiti.index = haiti.index.map(int) # let's change the index values of Haiti to type integer for plotting
haiti.plot(kind='line')

plt.title('Immigration from Haiti')
plt.ylabel('Number of immigrants')
plt.xlabel('Years')

plt.show() # need this line to show the updates made to the figure

Chúng ta có thể thấy rõ rằng số lượng người nhập cư từ Haiti đã tăng vọt từ năm 2010 khi Canada tăng cường tiếp nhận người tị nạn từ Haiti. Hãy biểu thị mức tăng đột biến này trong biểu đồ bằng phương thức `plt.text ()`.


In [ ]:
haiti.plot(kind='line')

plt.title('Immigration from Haiti')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')

# annotate the 2010 Earthquake.
# syntax: plt.text(x, y, label)
plt.text(2000, 6000, '2010 Earthquake') # see note below

plt.show()

Chỉ với một vài dòng code, bạn đã có thể nhanh chóng xác định và visualize mức tăng đột biến của việc nhập cư!

Một vài lưu ý về các giá trị x và y trong `plt.text (x, y, label)`:

```
 Since the x-axis (years) is type 'integer', we specified x as a year. The y axis (number of immigrants) is type 'integer', so we can just specify the value y = 6000.
```

```python
    plt.text(2000, 6000, '2010 Earthquake') # years stored as type int
```

```
If the years were stored as type 'string', we would need to specify x as the index position of the year. Eg 20th index is year 2000 since it is the 20th year with a base year of 1980.
```

```python
    plt.text(20, 6000, '2010 Earthquake') # years stored as type int
```

```
We will cover advanced annotation methods in later modules.
```


Chúng ta có thể dễ dàng thêm nhiều quốc gia hơn vào biểu đồ để so sánh luồng nhập cư từ các quốc gia khác nhau.

**Câu hỏi:** Hãy so sánh số lượng người nhập cư từ Ấn Độ và Trung Quốc từ năm 1980 đến năm 2013.


Bước 1: Lấy tập dữ liệu của Trung Quốc và Ấn Độ, đồng thời hiển thị khung dữ liệu.


In [ ]:
### type your answer here



Click đúp **vào đây** để xem lời giải.

<!-- The correct answer is:
df_CI = df_can.loc[['India', 'China'], years]
df_CI.head()
-->


Bước 2: Vẽ đồ thị. Chúng ta sẽ chỉ định rõ ràng biểu đồ dòng bằng cách truyền tham số `kind` vào `plot()`.


In [ ]:
### type your answer here



Click đúp **vào đây** để xem lời giải.

<!-- The correct answer is:
df_CI.plot(kind='line')
-->


Hình như có điều gì đó không ổn nhỉ...

Các bạn hãy nhớ rằng các biểu đồ _pandas_ sẽ có các chỉ mục trên trục x và và các cột ở dạng các dòng riêng lẻ trên trục y. Vì `df_CI` là một khung dữ liệu với chỉ mục là `country` và các cột là `years`, đầu tiên, chúng ta phải hoán vị khung dữ liệu bằng phương thức `transpose()` để hoán đổi hàng và các cột.


In [ ]:
df_CI = df_CI.transpose()
df_CI.head()

_pandas_ sẽ tự động vẽ biểu đồ 2 quốc gia trên cùng một biểu đồ. Hãy vẽ khung dữ liệu hoán vị mới. Hãy nhớ thêm tiêu đề cho biểu đồ và gắn nhãn cho các trục.


In [ ]:
### type your answer here





Click đúp **vào đây** để xem lời giải.

<!-- The correct answer is:
df_CI.index = df_CI.index.map(int) # let's change the index values of df_CI to type integer for plotting
df_CI.plot(kind='line')
-->

<!--
plt.title('Immigrants from China and India')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')
-->

<!--
plt.show()
-->


Từ biểu đồ trên, chúng ta có thể thấy rằng Trung Quốc và Ấn Độ có xu hướng nhập cư rất giống nhau qua các năm.


_Lưu ý_: Tại sao chúng ta lại không cần hoán vị khung dữ liệu của Haiti trước khi vẽ biểu đồ (như chúng ta đã làm với df_CI)?

Lý do là vì `haiti` là một chuỗi, chứ không phải là một khung dữ liệu, và với `haiti`, các năm là các chỉ mục như bên dưới.

```python
print(type(haiti))
print(haiti.head(5))
```

> class 'pandas.core.series.Series' <br>
> 1980    1666 <br>
> 1981    3692 <br>
> 1982    3498 <br>
> 1983    2860 <br>
> 1984    1418 <br>
> Name: Haiti, dtype: int64 <br>


Biểu đồ đường là một công cụ rất tiện lợi để hiển thị các biến phụ thuộc (dependent variable) so với các biến độc lập (independent variable). Tuy nhiên, một biểu đồ không nên có nhiều hơn 5 - 10 đường;  nếu không sẽ khiến biểu đồ khó hiểu.


**Câu hỏi:** So sánh xu hướng của 5 quốc gia có lượng nhập cư vào Canada nhiều nhất.


In [ ]:
### type your answer here





Click đúp **vào đây** để xem lời giải.

<!-- The correct answer is:
\\ # Step 1: Get the dataset. Recall that we created a Total column that calculates the cumulative immigration by country. \\ We will sort on this column to get our top 5 countries using pandas sort_values() method.
\\ inplace = True paramemter saves the changes to the original df_can dataframe
df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)
-->

<!--
# get the top 5 entries
df_top5 = df_can.head(5)
-->

<!--
# transpose the dataframe
df_top5 = df_top5[years].transpose()
-->

<!--
print(df_top5)
-->

<!--
\\ # Step 2: Plot the dataframe. To make the plot more readeable, we will change the size using the `figsize` parameter.
df_top5.index = df_top5.index.map(int) # let's change the index values of df_top5 to type integer for plotting
df_top5.plot(kind='line', figsize=(14, 8)) # pass a tuple (x, y) size
-->

<!--
plt.title('Immigration Trend of Top 5 Countries')
plt.ylabel('Number of Immigrants')
plt.xlabel('Years')
-->

<!--
plt.show()
-->


### Các biểu đồ khác

Xin chúc mừng! Bạn đã học cách xử lý dữ liệu bằng python và tạo biểu đồ đường bằng Matplotlib. Ngoài biểu đồ đường mặc định thì có nhiều kiểu vẽ biểu đồ khác, chúng ta có thể truy cập vào tất cả chúng bằng cách chuyển từ khóa `kind` thành `plot()`. Bên dưới là danh sách đầy đủ các biểu đồ có sẵn:

-   `bar` là biểu đồ thanh dọc
-   `barh` là biểu đồ thanh ngang
-   `hist` là biểu đồ tần suất
-   `box` là biểu đồ hộp
-   `kde` hoặc `density` là biểu đồ mật độ
-   `area` là biểu đồ khu vực
-   `pie` là biểu đồ tròn
-   `scatter` là biểu đồ phân tán
-   `hexbin` là biểu đồ hexbin (biểu đồ được chia thành các ngăn lục giác)
